### DISCUSSION TITLE
Angular libraries dependencies between each other

### DISCUSSION THREAD

**Discussion by sciborrudnicki** [nrwl/nx#7026] [Feature Requests]
Hi. How can I create reusable angular libraries that depend on each other? I did it already but it's very hard to maintain with `@angular/cli` and with `nx` it's impossible because of circular dependency error. I read also **To resolve circular dependencies:** [here](https://nx.dev/latest/angular/structure/dependency-graph) but it's not enough for me because my approach **works**. The first library uses some of the features from the second library, and the second library uses some of the features from the first library.  For example:

I've created library `callback` that has class and type.

```typescript
// callback
import { ValidationError } from 'error';
export type ResultCallback = () => any;
export class Callback {

    public create() {
        throw new ValidationError();
   }
}
```

Then i've created `error` library that has `ValidationError` class.

```typescript
// error
import { ResultCallback } from 'callback';
class ValidationError extends Error {
    public set(callback: ResultCallback) {}
}
```

Both libraries have peer dependencies set, for error, it's `callback` for the 'callback' it's `error`. I want to have everything about callback in the library `callback` and everything about an error in the `error` library. What is the best approach in this situation?

**Comment by vmasek** [comment]
To avoid circular dependencies you should extract pieces that are used by both and create the import circle to separate lib.

From the code example you provided, I'm not really sure whether is it the case where you extract the `Callback` and `Error` related code to separate lib or whether the logic is supposed to be coupled & thus moved together.

Also, as a general rule, I'd recommend setting up linting (`nx-enforce-module-boundaries`) as it helps a lot by preventing such architectural problems.
